File: sdks/javascript/src/resources/templates.ts
/**
 * Templates resource for managing email templates
 */

import type { Mailat } from '../client'
import type { Template, CreateTemplateOptions, UpdateTemplateOptions } from '../types'

export class TemplatesResource {
  constructor(private client: Mailat) {}

  /**
   * Create a new template
   *
   * @example
   * ```typescript
   * const template = await client.templates.create({
   *   name: 'Welcome Email',
   *   subject: 'Welcome to {{company_name}}, {{first_name}}!',
   *   htmlBody: '<h1>Welcome {{first_name}}!</h1><p>Thanks for joining us.</p>'
   * })
   * ```
   */
  async create(options: CreateTemplateOptions): Promise<Template> {
    return this.client.post<Template>('/templates', options)
  }

  /**
   * Get template by ID
   */
  async get(id: string): Promise<Template> {
    return this.client.get<Template>(`/templates/${id}`)
  }

  /**
   * Update a template
   */
  async update(id: string, options: UpdateTemplateOptions): Promise<Template> {
    return this.client.put<Template>(`/templates/${id}`, options)
  }

  /**
   * Delete a template
   */
  async delete(id: string): Promise<void> {
    await this.client.delete(`/templates/${id}`)
  }

  /**
   * List all templates
   */
  async list(options: {
    page?: number
    limit?: number
    search?: string
  } = {}): Promise<Template[]> {
    return this.client.get<Template[]>('/templates', {
      page: options.page,
      limit: options.limit,
      search: options.search
    })
  }

  /**
   * Preview template with sample data
   *
   * @example
   * ```typescript
   * const preview = await client.templates.preview('template-id', {
   *   first_name: 'John',
   *   company_name: 'Acme Inc'
   * })
   * console.log(preview.html) // Rendered HTML
   * console.log(preview.subject) // Rendered subject
   * ```
   */
  async preview(id: string, variables?: Record<string, unknown>): Promise<{
    html: string
    text?: string
    subject: string
  }> {
    return this.client.post(`/templates/${id}/preview`, { variables })
  }

  /**
   * Get template variables
   * Returns list of variable names used in the template
   */
  async getVariables(id: string): Promise<string[]> {
    const template = await this.get(id)
    return template.variables || []
  }

  /**
   * Validate template syntax
   * Checks for valid Handlebars/Mustache syntax
   */
  async validate(options: {
    htmlBody: string
    subject?: string
    textBody?: string
  }): Promise<{
    valid: boolean
    errors?: {
      field: string
      message: string
      line?: number
    }[]
    variables: string[]
  }> {
    return this.client.post('/templates/validate', options)
  }

  /**
   * Duplicate a template
   */
  async duplicate(id: string, name?: string): Promise<Template> {
    return this.client.post<Template>(`/templates/${id}/duplicate`, { name })
  }

  /**
   * Enable a template
   */
  async enable(id: string): Promise<Template> {
    return this.update(id, { isActive: true })
  }

  /**
   * Disable a template
   */
  async disable(id: string): Promise<Template> {
    return this.update(id, { isActive: false })
  }

  /**
   * Extract variables from template content
   * Utility function that parses Handlebars/Mustache variables
   */
  extractVariables(content: string): string[] {
    const regex = /\{\{([^}]+)\}\}/g
    const variables = new Set<string>()
    let match

    while ((match = regex.exec(content)) !== null) {
      // Clean up variable name (remove #, /, ^, etc.)
      const variable = match[1].trim().replace(/^[#/^]/, '').split(' ')[0]
      if (variable && !variable.startsWith('!')) {
        variables.add(variable)
      }
    }

    return Array.from(variables)
  }
}

File: sdks/python/mailat/resources/templates.py
"""Templates resource for managing email templates"""

from typing import TYPE_CHECKING, Any, Optional
import re
from ..types import Template

if TYPE_CHECKING:
    from ..client import Mailat


class TemplatesResource:
    """Resource for managing email templates."""

    def __init__(self, client: "Mailat"):
        self._client = client

    def create(
        self,
        name: str,
        subject: str,
        html_body: str,
        text_body: Optional[str] = None,
        description: Optional[str] = None,
    ) -> Template:
        """Create a new template."""
        payload = {
            "name": name,
            "subject": subject,
            "htmlBody": html_body,
        }
        if text_body:
            payload["textBody"] = text_body
        if description:
            payload["description"] = description

        data = self._client.post("/templates", json=payload)
        return Template(**data)

    def get(self, template_id: str) -> Template:
        """Get template by ID."""
        data = self._client.get(f"/templates/{template_id}")
        return Template(**data)

    def update(
        self,
        template_id: str,
        name: Optional[str] = None,
        subject: Optional[str] = None,
        html_body: Optional[str] = None,
        text_body: Optional[str] = None,
        description: Optional[str] = None,
        is_active: Optional[bool] = None,
    ) -> Template:
        """Update a template."""
        payload = {}
        if name:
            payload["name"] = name
        if subject:
            payload["subject"] = subject
        if html_body:
            payload["htmlBody"] = html_body
        if text_body:
            payload["textBody"] = text_body
        if description:
            payload["description"] = description
        if is_active is not None:
            payload["isActive"] = is_active

        data = self._client.put(f"/templates/{template_id}", json=payload)
        return Template(**data)

    def delete(self, template_id: str) -> None:
        """Delete a template."""
        self._client.delete(f"/templates/{template_id}")

    def list(
        self,
        page: int = 1,
        limit: int = 50,
        search: Optional[str] = None,
    ) -> list[Template]:
        """List all templates."""
        params = {"page": page, "limit": limit}
        if search:
            params["search"] = search

        data = self._client.get("/templates", params=params)
        return [Template(**t) for t in data]

    def preview(
        self,
        template_id: str,
        variables: Optional[dict[str, Any]] = None,
    ) -> dict[str, str]:
        """Preview template with sample data."""
        payload = {"variables": variables} if variables else {}
        return self._client.post(f"/templates/{template_id}/preview", json=payload)

    def get_variables(self, template_id: str) -> list[str]:
        """Get template variables."""
        template = self.get(template_id)
        return template.variables or []

    def validate(
        self,
        html_body: str,
        subject: Optional[str] = None,
        text_body: Optional[str] = None,
    ) -> dict[str, Any]:
        """Validate template syntax."""
        payload = {"htmlBody": html_body}
        if subject:
            payload["subject"] = subject
        if text_body:
            payload["textBody"] = text_body

        return self._client.post("/templates/validate", json=payload)

    def enable(self, template_id: str) -> Template:
        """Enable a template."""
        return self.update(template_id, is_active=True)

    def disable(self, template_id: str) -> Template:
        """Disable a template."""
        return self.update(template_id, is_active=False)

    @staticmethod
    def extract_variables(content: str) -> list[str]:
        """Extract variables from template content."""
        pattern = r"\{\{([^}]+)\}\}"
        matches = re.findall(pattern, content)
        variables = set()
        for match in matches:
            # Clean up variable name
            var = match.strip().lstrip("#/^").split()[0]
            if var and not var.startswith("!"):
                variables.add(var)
        return list(variables)
